frenchy64/clojurescript
# CLJS-2257: Split :var op, desugar dotted symbols and add tests+doc for analyzer

- :var/:binding/:local/:js-var
  - split :var op into :var/:binding/:local/:js-var
  - desugar dotted :var/:local symbols
    - ensure :context is correctly updated to avoid multiple `return`s
      - Tested: cljs.core-test/var-desugar-test
  - emit :local op if :js-shadowed-by-local
  - change :local to be #{:fn :letfn :let :arg ...}
  - add :arg-id
  - workaround for core.async's updating of :locals
    - see cljs.analyzer/handle-symbol-local
    - Test: cljs.analyzer-tests/test-locals-mapped-to-sym
- argument validation in 'var parsing
  - Tested: cljs.analyzer-tests/var-args-error-test
- :body? entries for synthetic `do` blocks
- Add analyzer Unit tests
  - cljs.analyzer-tests/analyze-ops
- Add AST documentation
  - based on tools.analyzer's, see ast-ref/ folder

========
Commits:
========

make :let/:loop/:letfn :binding ops

make :fn-method :binding ops

add :js-var and :local ops

use :local op if :js-shadowed-by-local

change :local to be #{:fn :letfn :let :arg ...}

merge :info into analyze-symbol output

add unit tests

desugar dotted symbols

don't use symbol namespaces to store :local

desugar :var

unit tests

argument validation in 'var parsing

bad (var sym) form tests

ast ref

add :body? labels

alpha

tweak doc

nicer diff

update doc

work around core.async's :locals extensions

update :context when desugaring dotted vars

add test for core.async workaround

remove quickref.html
